sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -229,42 +229,183 @@ CURLcode CurlSession::HttpRawSend()
   {
     // Try to fill internal buffer from socket.
     // If response is smaller than buffer, we will get back the size of the response
-    auto bufferSize = ReadSocketToBuffer(this->m_readBuffer, LibcurlReaderSize);
+    bufferSize = ReadSocketToBuffer(this->m_readBuffer, LibcurlReaderSize);
 
     // parse from buffer to create response
     auto bytesParsed = parser.Parse(this->m_readBuffer, (size_t)bufferSize);
     // if end of headers is reach before the end of response, that's where body start
-    if (bytesParsed < bufferSize)
+    if (bytesParsed + 2 < bufferSize)
     {
-      this->m_bodyStartInBuffer = bytesParsed + 1; // Set the start of body (skip \n)
+      this->m_bodyStartInBuffer = bytesParsed + 1; // Set the start of body (skip \r)
     }
   }
 
   this->m_response = parser.GetResponse();
+  this->m_innerBufferSize = static_cast<size_t>(bufferSize);
 
   // For Head request, set the length of body response to 0.
   if (this->m_request.GetMethod() == HttpMethod::Head)
   {
-    this->m_response->SetBodyStream(new CurlBodyStream(0, this));
+    this->m_response->SetBodyStream(std::make_unique<CurlBodyStream>(0, this));
     return CURLE_OK;
   }
 
   // TODO: tolower ContentLength
   auto headers = this->m_response->GetHeaders();
-  auto bodySize = atoi(headers.at("Content-Length").data());
-  // content-length is used later by session and session won't have access to the response any more
-  // (unique_ptr), so we save this value
-  this->m_contentLength = bodySize;
-  // Move session to live inside the stream from response.
-  this->m_response->SetBodyStream(new CurlBodyStream(bodySize, this));
 
-  return CURLE_OK;
+  auto isContentLengthHeaderInResponse = headers.find("Content-Length");
+  if (isContentLengthHeaderInResponse != headers.end())
+  {
+    // Response with Content-Length
+    auto bodySize = atoi(headers.at("Content-Length").data());
+    // content-length is used later by session and session won't have access to the response any
+    // more (unique_ptr), so we save this value
+    this->m_contentLength = bodySize;
+    // Move session to live inside the stream from response.
+    this->m_response->SetBodyStream(std::make_unique<CurlBodyStream>(bodySize, this));
+    return CURLE_OK;
+  }
+
+  auto isTransferEncodingHeaderInResponse = headers.find("Transfer-Encoding");
+  if (isTransferEncodingHeaderInResponse != headers.end())
+  {
+    auto headerValue = isTransferEncodingHeaderInResponse->second;
+    auto isChunked = headerValue.find("chunked");
+
+    if (isChunked != std::string::npos)
+    {
+      // set curl session to know response is chunked
+      // This will be used to remove chunked info while reading
+      this->m_chunkedBody = true;
+    }
+
+    // Response with no content-length.
+    this->m_response->SetBodyStream(std::make_unique<CurlBodyStream>(this));
+    return CURLE_OK;
+  }
+
+  // Response with no Content-Length and no Transfer-Encoding header. Throw
+  throw;
```

Comment:
This can happen, I met once just now.
We cannot just `throw;`, it will crash the whole program.

Reply:
it looks like in case of no Transfer-Encoding or Content-Length, we should just read until the connection is closed.

https://tools.ietf.org/html/rfc7230#section-3.3.3
> 7.  Otherwise, this is a response message without a declared message
       body length, so the message body length is determined by the
       number of octets received prior to the server closing the
       connection.

I going to remove the `throw` for now.

I need to improve the reading strategies to avoid code duplication,  I will make a new issue for it

Reply:
okay, just looked through the link. It seems that this HTTP1.1 is very complex and Azure Core team is re-implementating the HTTP stack from scratch.